### DISCUSSION TITLE
vkCreateRenderPass2 crashes [resolved]

### DISCUSSION THREAD

**Discussion by fhoenig** [KhronosGroup/MoltenVK#1778] [General]
I have a MoltenVK vulkan 1.2 instance and after the swapchain was created sucessfully I get the following problem:
What could this be? The device is valid and this works on Windows/NVIDIA.

`vkCreateRenderPass2` crashes with:
`Thread 1: EXC_BAD_ACCESS (code=1, address=0x0)`

```
VkRenderPassCreateInfo2	
-----------------------------
sType	VkStructureType	VK_STRUCTURE_TYPE_RENDER_PASS_CREATE_INFO_2
pNext	const void *	NULL	0x0000000000000000
flags	VkRenderPassCreateFlags	0
attachmentCount	uint32_t	1
pAttachments	const VkAttachmentDescription2 *	0x600003ef6140	0x0000600003ef6140
    sType	VkStructureType	VK_STRUCTURE_TYPE_ATTACHMENT_DESCRIPTION_2
    pNext	const void *	NULL	0x0000000000000000
    flags	VkAttachmentDescriptionFlags	0
    format	VkFormat	VK_FORMAT_B8G8R8A8_UNORM
    samples	VkSampleCountFlagBits	VK_SAMPLE_COUNT_1_BIT
    loadOp	VkAttachmentLoadOp	VK_ATTACHMENT_LOAD_OP_CLEAR
    storeOp	VkAttachmentStoreOp	VK_ATTACHMENT_STORE_OP_STORE
    stencilLoadOp	VkAttachmentLoadOp	VK_ATTACHMENT_LOAD_OP_DONT_CARE
    stencilStoreOp	VkAttachmentStoreOp	VK_ATTACHMENT_STORE_OP_DONT_CARE
    initialLayout	VkImageLayout	VK_IMAGE_LAYOUT_UNDEFINED
    finalLayout	VkImageLayout	VK_IMAGE_LAYOUT_PRESENT_SRC_KHR
subpassCount	uint32_t	1
pSubpasses	const VkSubpassDescription2 *	0x600000fb5e60	0x0000600000fb5e60
    sType	VkStructureType	VK_STRUCTURE_TYPE_SUBPASS_DESCRIPTION_2
    pNext	const void *	NULL	0x0000000000000000
    flags	VkSubpassDescriptionFlags	0
    pipelineBindPoint	VkPipelineBindPoint	VK_PIPELINE_BIND_POINT_GRAPHICS
    viewMask	uint32_t	0
    inputAttachmentCount	uint32_t	0
    pInputAttachments	const VkAttachmentReference2 *	NULL	0x0000000000000000
    colorAttachmentCount	uint32_t	1
    pColorAttachments	const VkAttachmentReference2 *	0x7fc8008af800	0x00007fc8008af800
        sType	VkStructureType	VK_STRUCTURE_TYPE_ATTACHMENT_REFERENCE_2
        pNext	const void *	NULL	0x0000000000000000
        attachment	uint32_t	0
        layout	VkImageLayout	VK_IMAGE_LAYOUT_COLOR_ATTACHMENT_OPTIMAL
        aspectMask	VkImageAspectFlags	0
pResolveAttachments	const VkAttachmentReference2 *	NULL	0x0000000000000000
pDepthStencilAttachment	const VkAttachmentReference2 *	NULL	0x0000000000000000
preserveAttachmentCount	uint32_t	0
pPreserveAttachments	const uint32_t *	NULL	0x0000000000000000
dependencyCount	uint32_t	0
pDependencies	const VkSubpassDependency2 *	NULL	0x0000000000000000
correlatedViewMaskCount	uint32_t	0
pCorrelatedViewMasks	const uint32_t *	NULL	0x0000000000000000
```

**Comment by billhollings** [comment]
Do you have more info? A stack trace, perhaps, showing what's happening when the error is triggered?

**Reply by fhoenig** [reply]
Looks like it works with `vkCreateRenderPass` but then I cannot use a separate depth/stencil attachment. I only need one float32 depth buffer without a stencil. 

Regarding the stack trace, I don't get anything deeper than that call crashing. What would it take to get a debug version of this setup? I use the lunarg SDK installed into /usr/local and link to the vulkan lib.

**Comment by billhollings** [comment]
> Regarding the stack trace, I don't get anything deeper than that call crashing. What would it take to get a debug version of this setup? I use the lunarg SDK installed into /usr/local and link to the vulkan lib.

Often a stack track is logged to your console on a crash.

You can create a Debug version of MoltenVK by following the [build instructions](https://github.com/KhronosGroup/MoltenVK#building) in the `README.md` document from the MoltenVK repository. Then either copy `MoltenVK/Package/Debug/MoltenVK/dylib/macOS/libMoltenVK.dylib` to `/usr/local`, or create a symlink in `/usr/local` to point to it.

**Reply by fhoenig** [reply]
There is nothing in stdout or stderr. 
Just for clarification: for `vkCreateRenderPass2` I do not need to load the function separately? The loader should do it?

**Reply by cdavis5e** [reply]
> There is nothing in stdout or stderr.

He may be referring to the system log viewer app at `/Applications/Utilities/Console.app`. The crash dialog that normally pops up should also have it.

> Just for clarification: for `vkCreateRenderPass2` I do not need to load the function separately? The loader should do it?

Yes. I believe only core entry points are exported directly; extension entry points need to be loaded with `vkGetInstanceProcAddr()` or `vkGetDeviceProcAddr()`. `VK_KHR_create_renderpass2` was promoted to core in 1.2.

**Reply by fhoenig** [reply]
Happens exactly on the `vkCreateRenderPass2` call.

```
Crashed Thread:        0  Dispatch queue: com.apple.main-thread

Exception Type:        EXC_BAD_ACCESS (SIGSEGV)
Exception Codes:       KERN_INVALID_ADDRESS at 0x0000000000000000
Exception Codes:       0x0000000000000001, 0x0000000000000000
Exception Note:        EXC_CORPSE_NOTIFY

Termination Reason:    Namespace SIGNAL, Code 11 Segmentation fault: 11
Terminating Process:   exc handler [11166]

VM Region Info: 0 is not in any region.  Bytes before following region: 4534235136
      REGION TYPE                    START - END         [ VSIZE] PRT/MAX SHRMOD  REGION DETAIL
      UNUSED SPACE AT START
--->  
      __TEXT                      10e42f000-11055f000    [ 33.2M] r-x/r-x SM=COW  ...MacOS/Unbound

Thread 0 Crashed::  Dispatch queue: com.apple.main-thread
0   ???                           	               0x0 ???
1   Unbound                       	       0x10fde0b6c Gfx::VK::Device::createRenderPass(Gfx::RenderPassDescriptor) + 1900 (Device.cpp:1120)
2   Unbound                       	       0x10e71215e Renderer::init(SDL_Window*, std::__1::shared_ptr<Gfx::IDevice>) + 1518 (Renderer.cpp:440)
3   Unbound                       	       0x10e624814 Engine::init() + 628 (Context.cpp:265)
4   Unbound                       	       0x10e62587e Engine::run() + 62 (Context.cpp:473)
5   Unbound                       	       0x10e70c4b7 main + 39 (PlayEngine.cpp:17)
6   dyld                          	       0x11f0bb52e start + 462
```

**Reply by fhoenig** [reply]
Also nothing in the logs via Console.app

**Reply by fhoenig** [reply]
Both `vkGetDeviceProcAddr(m_device, "vkCreateRenderPass2");` 
and `vkGetDeviceProcAddr(m_device, "vkCreateRenderPass2KHR");` 
return `NULL`

However `vkGetDeviceProcAddr(m_device, "vkCreateRenderPass");` works.

**Reply by cdavis5e** [reply]
What version of MoltenVK are you using? I'm 100% certain that `VK_KHR_create_renderpass2` is supported, because I implemented it myself.

**Reply by fhoenig** [reply]
The latest from Lunar 1.3.231.1

**Comment by fhoenig** [comment]
Solved. 

Looks like the python installer script had failed to copy the libs from the most recent sdk.
I manually copied them into /usr/local/ and that made it work.

**Reply by billhollings** [reply]
Glad you got it sorted. Thanks for letting us know!

**Reply by fhoenig** [reply]
Thanks for reading.
